CD Syscall Types Update

QuzarDC opened the pull request:
This is in the works to get in the typing and cleanup suggestions from #619 and #736 with a different scope.

The key purpose is to add strongly defined types for the various returns and parameters of the functions for interacting with the GDROM drive, replacing the many sets of defines, generic types, and untyped structs.

Changes:
- [x] Redefine each group of CD defines to an enum, per #619. (The only one left out is the unused group of ATA Status defines @DC-SWAT perhaps you have more info on them since you added them? Are these returned by one of the syscalls? Maybe they're the 4th param of Check Command?)
- [x] Move syscall related define groups to syscalls.h
- [x] Rename some of the define group values to better match our general naming scheme (`CD_ERR_OK` rather than just `ERR_OK`).
- [ ] Cleanups to structure (some due to the typing, some due to the scoped mutex or syscalls updates)
- [ ] Create types for used param structs.
- [ ] Clean up the documentation, ensuring there are good cross-referencces between the data types and functions that they get used by.

Possible further work:
- [ ] Add abort command (per #736) with some refactoring of the command issuance.
- [ ] Separate out gd syscall things into a separate header from the others.
- [ ] Fully type each of the syscall's param structs.

Still need a bit more work for what I was sure to add, but wanted to make sure to make it public now to get feedback/help with the documentation, naming, and just general review for mistakes.

DC-SWAT commented on kernel/arch/dreamcast/include/dc/syscalls.h:
```diff
@@ -258,18 +496,20 @@ int syscall_gdrom_check_command(uint32_t id, int32_t status[4]);
 void syscall_gdrom_exec_server(void);
 
-/** \brief   Abort a queued GDROM command.
+/** \brief      Abort a queued GDROM command.
```
There are no queues, you can only request one command.

DC-SWAT commented on kernel/arch/dreamcast/include/dc/syscalls.h:
```diff
@@ -246,9 +483,10 @@ uint32_t syscall_gdrom_send_command(uint32_t cmd, void *params);
+cmd_cmd_chk_t syscall_gdrom_check_command(gdc_cmd_id_t id, int32_t status[4]);
 
-/** \brief   Process queued GDROM commands.
+/** \brief      Process queued GDROM commands.
```
There are no queues, you can only request one command.

DC-SWAT reviewed (commented):
Are you sure you want to break the API that much?

DC-SWAT commented:
> This is in the works to get in the typing and cleanup suggestions from #619 and #736 with a different scope.
> 
> The key purpose is to add strongly defined types for the various returns and parameters of the functions for interacting with the GDROM drive, replacing the many sets of defines, generic types, and untyped structs.
> 
> Changes:
> 
> * [x]  Redefine each group of CD defines to an enum, per [Clean up CD-ROM code #619](https://github.com/KallistiOS/KallistiOS/pull/619). (The only one left out is the unused group of ATA Status defines @DC-SWAT perhaps you have more info on them since you added them? Are these returned by one of the syscalls? Maybe they're the 4th param of Check Command?)

Yeah this is it. 4th param in command status.

QuzarDC commented:
> Are you sure you want to break the API that much?

Could you please elaborate? This shouldn't break much of anything. The renaming of the error codes and `NO_ACTIVE` to `INACTIVE` is possibly the worst, but those weren't mapped to any external use. You might get warnings but shouldn't get errors from passing the old plain types in or assigning the return of a function to an `int` rather than a `cd_cmd_ret_t`. Maybe I'm horribly wrong, or not considering some uses or settings?

Now that you mention it though, I could add a compatibility set for the changed define names to ensure just a warning for those too.

DC-SWAT commented:
> > Are you sure you want to break the API that much?
> 
> Could you please elaborate? This shouldn't break much of anything. The renaming of the error codes and `NO_ACTIVE` to `INACTIVE` is possibly the worst, but those weren't mapped to any external use. You might get warnings but shouldn't get errors from passing the old plain types in or assigning the return of a function to an `int` rather than a `cd_cmd_ret_t`. Maybe I'm horribly wrong, or not considering some uses or settings?
> 
> Now that you mention it though, I could add a compatibility set for the changed define names to ensure just a warning for those too.

You renamed the error codes that are used externally.
I mean for example ERR_SYS -> CD_ERR_SYS.
I have 2 applications in DreamsShell that use the сdrom api and I can say for sure that they will break. It is not a big problem of course, to rename constants, but nevertheless. I think that all programs that used сdrom api will break.
It just seemed to me that we were trying not to break backward compatibility all time. But here it was done without any particular need.

QuzarDC commented:
> > > Are you sure you want to break the API that much?
> > 
> > 
> > Could you please elaborate? This shouldn't break much of anything. The renaming of the error codes and `NO_ACTIVE` to `INACTIVE` is possibly the worst, but those weren't mapped to any external use. You might get warnings but shouldn't get errors from passing the old plain types in or assigning the return of a function to an `int` rather than a `cd_cmd_ret_t`. Maybe I'm horribly wrong, or not considering some uses or settings?
> > Now that you mention it though, I could add a compatibility set for the changed define names to ensure just a warning for those too.
> 
> You renamed the error codes that are used externally. I mean for example ERR_SYS -> CD_ERR_SYS. I have 2 applications in DreamsShell that use the сdrom api and I can say for sure that they will break. It is not a big problem of course, to rename constants, but nevertheless. I think that all programs that used сdrom api will break. It just seemed to me that we were trying not to break backward compatibility all time. But here it was done without any particular need. More precisely, of course it’s better to name it as you did, but it breaks.

Ok, that should be the simplest thing to provide a compatibility layer for, as I did with `CDROM_TOC`. Then remove one version out (similar to removing the old `change_dataype` typo now). The rest of the type changes wouldn't break anything for you right? If it's just the err defines then we should be good :)

DC-SWAT commented:
> > > > Are you sure you want to break the API that much?
> 
> > > 
> 
> > > 
> 
> > > Could you please elaborate? This shouldn't break much of anything. The renaming of the error codes and `NO_ACTIVE` to `INACTIVE` is possibly the worst, but those weren't mapped to any external use. You might get warnings but shouldn't get errors from passing the old plain types in or assigning the return of a function to an `int` rather than a `cd_cmd_ret_t`. Maybe I'm horribly wrong, or not considering some uses or settings?
> 
> > > Now that you mention it though, I could add a compatibility set for the changed define names to ensure just a warning for those too.
> 
> > 
> 
> > You renamed the error codes that are used externally. I mean for example ERR_SYS -> CD_ERR_SYS. I have 2 applications in DreamsShell that use the сdrom api and I can say for sure that they will break. It is not a big problem of course, to rename constants, but nevertheless. I think that all programs that used сdrom api will break. It just seemed to me that we were trying not to break backward compatibility all time. But here it was done without any particular need. More precisely, of course it’s better to name it as you did, but it breaks.
> 
> 
> 
> Ok, that should be the simplest thing to provide a compatibility layer for, as I did with `CDROM_TOC`. Then remove one version out (similar to removing the old `change_dataype` typo now). The rest of the type changes wouldn't break anything for you right? If it's just the err defines then we should be good :) 

Yeah, only error codes and TOC.

QuzarDC commented:
@DC-SWAT , I've removed the reference to queuing in favor of 'request' as described. I also incorporated the abort command and refactor from #736 to try to reduce the overhead of requesting further changes from you. I know that this will require some rewriting (aside from what is needed for this whole PR) but I hope the changes made sense.

DC-SWAT commented on kernel/arch/dreamcast/hardware/syscalls.c:
```diff
@@ -204,29 +204,29 @@ void syscall_gdrom_dma_callback(uintptr_t callback, void *param) {
 }
 
-int syscall_gdrom_pio_transfer(uint32_t id, const int32_t params[2]) {
+int syscall_gdrom_pio_transfer(gdc_cmd_hnd_t hnd, const int32_t params[2]) {
```
The same here.

QuzarDC commented on kernel/arch/dreamcast/hardware/syscalls.c:
```diff
@@ -204,29 +204,29 @@ void syscall_gdrom_dma_callback(uintptr_t callback, void *param) {
 }
 
-int syscall_gdrom_dma_transfer(uint32_t id, const int32_t params[2]) {
+int syscall_gdrom_dma_transfer(gdc_cmd_hnd_t hnd, const int32_t params[2]) {
```
Thank you! Updated.

andressbarajas commented on kernel/arch/dreamcast/include/dc/syscalls.h:
```diff
@@ -183,42 +182,324 @@ int syscall_flashrom_write(uint32_t pos, const void *src, size_t n);
+typedef struct cd_read_params {
+    uint32_t    start_sec;  /* Starting sector */
+    size_t      num_sec;    /* Number of sectors */
+    void       *buffer;     /* Output buffer */
```
```suggestion
    void        *buffer;     /* Output buffer */
```

andressbarajas commented on kernel/arch/dreamcast/include/dc/syscalls.h:
```diff
@@ -183,42 +182,324 @@ int syscall_flashrom_write(uint32_t pos, const void *src, size_t n);
+typedef struct cd_cmd_getscd_params {
+    cd_sub_type_t   which;
+    size_t          buflen;
+    void           *buffer;
```
```suggestion
    void            *buffer;
```

andressbarajas commented on kernel/arch/dreamcast/include/dc/syscalls.h:
```diff
@@ -292,12 +630,23 @@ int syscall_gdrom_sector_mode(uint32_t mode[4]);
+    These are parameters passed to the syscall_gdrom_*_transfer syscalls.
+*/
+typedef struct cd_transfer_params {
+    void*   addr;       /**< \brief Destination address of transfer */
```
```suggestion
    void    *addr;       /**< \brief Destination address of transfer */
```

DC-SWAT commented:
Are you going to finish the last 2 points in this PR?
It needs to be merged. The only request is to return the abortion back to the cdrom driver.

QuzarDC commented:
I've gone ahead and rebuilt this PR. The commits should now all be quite atomic and cleanly separated. The big thing that I left out was implementing the same for the error codes. Those are going to require a substantial amount of riskier changes as we are currently mixing and matching the different error code types within cdrom.c as well as hijacking some for our own errors. So it seems appropriate to have that as a separate PR that isn't as big.

QuzarDC commented on kernel/arch/dreamcast/include/dc/syscalls.h:
```diff
@@ -183,42 +183,244 @@ int syscall_flashrom_write(uint32_t pos, const void *src, size_t n);
+    CD_CMD_DMAREAD           = 17,  /**< \brief Read via DMA */
+    CD_CMD_GETTOC            = 18,  /**< \brief Read TOC */
+    CD_CMD_GETTOC2           = 19,  /**< \brief Read TOC */
+    CD_CMD_PLAY              = 20,  /**< \brief Play track */
```
Oh! Of course, great idea. Do you know if there's similar kind of info on the difference between GETTOC/GETTOC2?

DC-SWAT replied:
All games use only `GETTOC2`, so I don't even emulate `GETTOC`. But I assume that this variant does not have a division into low density and high density areas and contains tracks only for low density.

QuzarDC commented:
Last two force-pushes were simply to ensure there were still no conflicts after other merges and we're up to date. No changes to the commits.

QuzarDC commented:
Force-pushed to rebase onto master after #1172 .

QuzarDC commented:
Had to rebase to account for the change in #1217 to use a semaphore rather than mutex for dma.

darcagn commented:
Seems to be working OK now. This is a big one; hopefully we didn't miss anything...
